### PR TITLE
Scope build debugging trace file to the build operation

### PR DESCRIPTION
**Pull request by jakepetroules**
Clients who cache and re-use a BuildSystem instance to serve multiple subsequent build operations are unable to get a trace file for multiple builds.

This is because the trace file pointer is scoped to and kept open for the lifetime of the BuildSystem instance, rather than being scoped to the build operation.

This change scopes the trace file writing to the build operation. This means that multiple build operations by the same BuildSystem will no longer be able to "append" to the same trace file (e.g. having multiple build-started/build-ended pairs), but this responsibility can be shifted to clients who can preserve the file between builds if desired.

rdar://47133775

### CODE REVIEW

**Comment by jakepetroules** [comment]
@swift-ci please smoke test

File: lib/Core/BuildEngine.cpp [hunk 1]
```diff
@@ -1534,6 +1532,23 @@ class BuildEngineImpl : public BuildDBDelegate {
...
+
+      this->trace = std::move(trace);
+    }
+
+    llbuild_defer {
+      std::string error;
+      if (!trace->close(&error)) {
+        delegate.error(error);
+      }
+    };
```
Comment (dmbryson) [inline lib/Core/BuildEngine.cpp:1550]: Should we be clearing out/storing nullptr to `this->trace` here too?

**Review by dmbryson — APPROVED** [review]

File: lib/Core/BuildEngine.cpp (hunk 1 above)
Comment (jakepetroules) [reply lib/Core/BuildEngine.cpp:1552]: Good point; done!

**Comment by jakepetroules** [comment]
@swift-ci please smoke test

**Comment by jakepetroules** [comment]
@swift-ci please smoke test

**Pull request merged**